### PR TITLE
Bump otelgrpc to fix CVE issue

### PR DESCRIPTION
**Pull request by wingyplus**
Found in the Trivy scan in #6099. Bump `otelgrpc` library to 0.46.0 and cleanup dependencies with `go mod tidy`.

### CODE REVIEW

**Comment by wingyplus** [comment]
Not a good way to fix. :(

**Comment by wingyplus** [comment]
Compared to https://github.com/dagger/dagger/actions/runs/6851646021/job/18628377264?pr=6099#step:7:310, we have only 1 `otelgrpc` issue related to buildkit binary. The pending issue is lint, it fails because of the function deprecation in `otelgrpc`.

I'm not sure about other issues in Trivy but it seems we need to update dependencies in the Alpine during build.

**Comment by sipsma** [comment]
Thanks for looking into this @wingyplus! Your changes look good so far.

One of the remaining issues is the `buildctl` binary that we have to include in the engine image. We ran into similar issues w/ it and needing to upgrade vulnerable deps recently... I'm thinking it may be time to just squash this by removing the dependency on `buildctl` entirely.

I will give that a quick shot and push to this PR if successful.

**Comment by sipsma** [comment]
Fixed the `buildctl` issue here: https://github.com/dagger/dagger/pull/6100/commits/b67fe73dc11c3624286aceea3837fabc38515804

Basically, we just don't use buildctl anymore. The only reason it was in our engine image was for the `dial-stdio` subcommand, so I just extracted the relevant parts of that out (it's a very small chunk of code) and put it in our repo.

---

The other issue was due to caching of our `apk add` command resulting in us getting versions of alpine libs that were fixed a few days ago; fixed that here:  https://github.com/dagger/dagger/pull/6100/commits/44def4e31413d0e6c7c65d9a0ea5f0ce18b4357f


---

cc @jedevc @vito this now has meaningful changes outside go.mod, PTAL

**Review by jedevc — APPROVED** [review]

File: internal/mage/util/engine.go [hunk 1]
```diff
@@ -213,24 +215,33 @@ func devEngineContainer(c *dagger.Client, arch string, version string, opts ...D
 
 	container := c.Container(dagger.ContainerOpts{Platform: dagger.Platform("linux/" + arch)}).
 		From("alpine:"+alpineVersion).
+		// NOTE: wrapping the apk installs with this time based env ensures that the cache is invalidated
+		// once-per day. This is a very unfortunate workaround for the poor caching "apk add" as an exec
+		// gives us.
+		// Fortunately, better approaches are on the horizon w/ Zenith, for which there are already apk
+		// modules that fix this problem and always result in the latest apk packages for the given alpine
+		// version being used (with optimal caching).
+		WithEnvVariable("DAGGER_APK_CACHE_BUSTER", fmt.Sprintf("%d", time.Now().Unix()/86400)).
```
Comment (vito) [inline internal/mage/util/engine.go:224]: nit: using `(time.Time).Truncate` makes this a bit easier to grok

```suggestion
		WithEnvVariable("DAGGER_APK_CACHE_BUSTER", fmt.Sprintf("%d", time.Now().Truncate(24 * time.Hour).Unix())).
```

or:

```suggestion
		WithEnvVariable("DAGGER_APK_CACHE_BUSTER", time.Now().Truncate(24 * time.Hour).String()).
```

**Review by vito — APPROVED** [review]

File: internal/mage/util/engine.go (hunk 1 above)
Comment (sipsma) [reply internal/mage/util/engine.go]: Nice, I missed that `Truncate` existed! Updated

**Pull request merged**